reading/queue.livemd

```diff
@@ -0,0 +1,315 @@
+  end
+
+  def empty?(queue) do
+    :your_code_here
+  end
+end
+```
+
+Continuing, how would you *enqueue* an item -- let's say `:bob` -- to an empty Queue?
+
+```elixir
+defmodule Queue do
+  def new() do
+    :your_code_here
+  end
+
+  def empty?(queue) do
+    :your_code_here
+  end
+
+  def enqueue(queue, item) do
+    :your_code_here
+  end
+end
+```
+
+Continuing, it might be nice to know how many items are in the Queue.
+
+```elixir
+defmodule Queue do
+  def new() do
+    :your_code_here
+  end
+
+  def empty?(queue) do
+    :your_code_here
+  end
+
+  def enqueue(queue, item) do
+    :your_code_here
+  end
+
+  def size(queue) do
+    :your_code_here
+  end
+end
+```
+
+
+Continuing, how would you *dequeue* an item from our non-empty
+Queue? The *dequeue*d item's value should be `:bob`. Dont forget
+to return the new queue (without `:bob`) also!
+
+```elixir
+defmodule Queue do
+  def new() do
+    :your_code_here
+  end
+
+  def empty?(queue) do
+    :your_code_here
+  end
+
+  def enqueue(queue, item) do
+    :your_code_here
+  end
+
+  def size(queue) do
+    :your_code_here
+  end
+
+  def dequeue(queue) do
+    :your_code_here
+  end
+end
+```
+
+### Your Turn
+
+Imagine people forming a line, arriving and leaving from it.
+
+Using your Queue module above, create a queue where the following people
+arrived in this order:
+
+* Soren
+* Ram
+* Cindy
+* Sam
+
+```mermaid
+flowchart
+  subgraph Waiting
+    direction LR
+    Sam[Sam] --> Cindy[Cindy] --> Ram[Ram] --> Soren[Soren]
+  end
+```
+
+<details style="background-color: lightgreen; padding: 1rem; margin: 1rem 0;">
+<summary>Example solution</summary>
+
+```elixir
+defmodule Line do
+  def arrivals(people) do
+    Enum.reduce(people, Queue.new(), fn arrival, queue ->
+      Queue.enqueue(queue, arrival)
+    end)
+  end
+end
+
+people = ["Soren", "Ram", "Cindy", "Sam"]
+line = Line.arrivals(people)
+IO.inspect(line, label: "In line is")
+```
+
+</details>
+
+Enter your solution below.
+
+```elixir
+defmodule Line do
+  def arrivals(people) do
+    :your_code_here
+  end
+end
+
+people = []
+line = Line.arrivals(people)
+```
+
+Continuing, update your Line so that it can serve some number of people
+from the line. Have it serve the first 2 people from the line:
+
+  * Who gets served?
+  * Who remains in the line?
+
+<details style="background-color: lightgreen; padding: 1rem; margin: 1rem 0;">
+<summary>Example solution</summary>
+
+```elixir
+defmodule Line do
+  def arrivals(people) do
+    Enum.reduce(people, Queue.new(), fn arrival, queue ->
+      Queue.enqueue(queue, arrival)
+    end)
+  end
+
+  def serve(line, number) do
+    case number do
+      0 -> line
+      _ ->
+        {first, line} = Queue.dequeue(line)
+        IO.inspect(first, label: "Just served")
+        serve(line, number-1)
+    end
+  end
+end
+
+line = Line.serve(line, 2)
+IO.inspect(line, label: "Remaining in line is")
+```
+
+</details>
+
+Enter your solution below.
+
+```elixir
+:your_code_here
```

Comment:
Often I'll use a single documented module, then say: "Implement the `Line` module as documented" to declare the start of the exercise portion.

Thoughts?

Reply:
ah! i think i know what you are saying .. i whip one up!